pkg/asset/upgrade/bundle.go

```diff
@@ -0,0 +1,489 @@
+package upgrade
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/bombsimon/logrusr/v4"
+	"github.com/distribution/distribution/v3/reference"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"github.com/openshift/appliance/pkg/asset/config"
+	"github.com/openshift/appliance/pkg/jq"
+	"github.com/openshift/appliance/pkg/registry"
+	"github.com/openshift/installer/pkg/asset"
+)
+
+// Bundle generates the upgrade bundle.
+type Bundle struct {
+	Tar    *asset.File
+	Digest *asset.File
+}
+
+var _ asset.Asset = (*Bundle)(nil)
+
+// Name returns a human friendly name for the asset.
+func (b *Bundle) Name() string {
+	return "Upgrade bundle"
+}
+
+// Dependencies returns all the dependencies directly needed to generate the asset.
+func (b *Bundle) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		&config.EnvConfig{},
+		&config.ApplianceConfig{},
+	}
+}
+
+// Generate generates the upgrade bundle.
+func (b *Bundle) Generate(dependencies asset.Parents) error {
+	// Get the logger:
+	logger := logrus.StandardLogger()
+
+	// Get the dependencies:
+	envCfg := &config.EnvConfig{}
+	appCfg := &config.ApplianceConfig{}
+	dependencies.Get(envCfg, appCfg)
+
+	// Verify the pieces of the configuration that we need:
+	if appCfg.Config.OcpRelease.Version == "" {
+		return errors.New("version is mandatory")
+	}
+	if appCfg.Config.OcpRelease.CpuArchitecture == nil || *appCfg.Config.OcpRelease.CpuArchitecture == "" {
+		return errors.New("architecture is mandatory")
+	}
+	if appCfg.Config.PullSecret == "" {
+		return errors.New("pull secret is mandatory")
+	}
+
+	// Create the JSON query tool:
+	jqTool, err := jq.NewQueryTool().
+		SetLogger(logrusr.New(logger)).
+		Build()
+	if err != nil {
+		return err
+	}
+
+	// Create the temporary directory for the bundle contents:
+	tmpDir := filepath.Join(envCfg.TempDir, "upgrade")
+	err = os.Mkdir(tmpDir, 0700)
+	if errors.Is(err, os.ErrExist) {
+		err = nil
```

Comment:
We can ask the user to run 'clean' command first in the case, as done in build: https://github.com/openshift/appliance/blob/ae232a33d1c23b938dbff0b15ced32b9e88525d2/cmd/build.go#L45

Just need to add 'upgrade' dir to the clean command (so worth extracting to a constant).